leader/backoff.go

```diff
@@ -1,52 +1,94 @@
+	return &CountingBackOff{
+		backoff: backOff,
+	}
+}
+
+// CountingBackOff is an exponential backoff that
+// counts the number of backoff steps
+type CountingBackOff struct {
+	backoff backoff.BackOff
 
-	tries   int
-	elapsed time.Duration
+	tries int
 }
 
 // NextBackOff returns the duration of the next backoff interval
-func (f *FreebieExponentialBackOff) NextBackOff() time.Duration {
+func (f *CountingBackOff) NextBackOff() time.Duration {
 	f.tries++
-	if f.tries == 1 {
-		return time.Duration(0)
-	} else {
-		delay := float64(f.InitialInterval) * math.Pow(f.Multiplier, float64(f.tries-2))
-		jitter := (rand.Float64() - 0.5) * f.RandomizationFactor * float64(f.InitialInterval)
-		delay += jitter
-		if delay >= float64(f.MaxInterval) {
-			delay = float64(f.MaxInterval)
-		}
-		f.elapsed += time.Duration(delay)
-		if f.elapsed > f.MaxElapsedTime {
-			return backoff.Stop
-		}
-		return time.Duration(delay)
-	}
+	return f.backoff.NextBackOff()
 }
 
-// Reset resets the number of tries on this backoff counter to zero
-func (f *FreebieExponentialBackOff) Reset() {
+// Reset resets both the number of steps and the backoff interval to zero
+func (f *CountingBackOff) Reset() {
 	f.tries = 0
-	f.elapsed = time.Duration(0)
+	f.backoff.Reset()
 }
 
-// CurrentTries returns the number of attempts on this backoff counter
-func (f *FreebieExponentialBackOff) CurrentTries() int {
+// NumTries returns the number of steps taken along this backoff interval
+func (f *CountingBackOff) NumTries() int {
 	return f.tries
 }
+
+// NewFlippingBackOff returns a new instance of the FlippingBackOff
+// using regular and failing as backoff implementations
+func NewFlippingBackOff(regular, failing backoff.BackOff) *FlippingBackOff {
+	return &FlippingBackOff{
+		regular: regular,
+		failing: failing,
+	}
+}
+
+// Failing resets the failing state to failing.
+// If failing == false, the failing backoff interval is reset.
+func (r *FlippingBackOff) Failing(failing bool) {
+	r.isFailing = failing
+	if !failing {
+		r.failing.Reset()
+	}
+}
+
+// NextBackOff returns the duration of the next backoff interval
+func (r *FlippingBackOff) NextBackOff() time.Duration {
+	if r.isFailing {
+		return r.failing.NextBackOff()
+	} else {
+		return r.regular.NextBackOff()
+	}
+}
+
+// Reset resets the underlying backoff intervals
+func (r *FlippingBackOff) Reset() {
+	r.regular.Reset()
+	r.failing.Reset()
+}
+
+// FlippingBackOff provides a backoff using two backoff implementations
+// it alternates between.
+// First implementation is used to provide regular notifications,
+// while the second implementation is used to handle errors.
+//
+// This can be used in conjunction with the backoff.Ticker to provide a custom
+// loop that can dynamically switch between backoff implementations
+// depending on state of an operation (e.g. healthy vs experiencing transient errors).
+type FlippingBackOff struct {
+	regular, failing backoff.BackOff
```

Comment:
I would split these on two lines for clarity, and added comments on each (even though they're not exported). For example, it is not very clear to me what "regular" backoff is for. Is this basically for repeating function call when it returns nil error?